include/mbgl/style/filter.hpp

```diff
@@ -11,270 +11,28 @@
-public:
-    FeatureType value;
-
-    friend bool operator==(const TypeNotEqualsFilter& lhs, const TypeNotEqualsFilter& rhs) {
-        return lhs.value == rhs.value;
-    }
-};
-
-class TypeInFilter {
-public:
-    std::vector<FeatureType> values;
-
-    friend bool operator==(const TypeInFilter& lhs, const TypeInFilter& rhs) {
-        return lhs.values == rhs.values;
-    }
-};
-
-class TypeNotInFilter {
-public:
-    std::vector<FeatureType> values;
-
-    friend bool operator==(const TypeNotInFilter& lhs, const TypeNotInFilter& rhs) {
-        return lhs.values == rhs.values;
-    }
-};
-
-
-class IdentifierEqualsFilter {
-public:
-    FeatureIdentifier value;
-
-    friend bool operator==(const IdentifierEqualsFilter& lhs, const IdentifierEqualsFilter& rhs) {
-        return lhs.value == rhs.value;
-    }
-};
-
-class IdentifierNotEqualsFilter {
-public:
-    FeatureIdentifier value;
-
-    friend bool operator==(const IdentifierNotEqualsFilter& lhs, const IdentifierNotEqualsFilter& rhs) {
-        return lhs.value == rhs.value;
-    }
-};
-
-class IdentifierInFilter {
-public:
-    std::vector<FeatureIdentifier> values;
-
-    friend bool operator==(const IdentifierInFilter& lhs, const IdentifierInFilter& rhs) {
-        return lhs.values == rhs.values;
-    }
-};
-
-class IdentifierNotInFilter {
-public:
-    std::vector<FeatureIdentifier> values;
-
-    friend bool operator==(const IdentifierNotInFilter& lhs, const IdentifierNotInFilter& rhs) {
-        return lhs.values == rhs.values;
-    }
-};
-
-class HasIdentifierFilter {
-public:
-    friend bool operator==(const HasIdentifierFilter&, const HasIdentifierFilter&) {
-        return true;
-    }
-};
-
-class NotHasIdentifierFilter {
-public:
-    friend bool operator==(const NotHasIdentifierFilter&, const NotHasIdentifierFilter&) {
-        return true;
-    }
-};
     
-class ExpressionFilter {
+class Filter {
 public:
     std::shared_ptr<const expression::Expression> expression;
     
-    friend bool operator==(const ExpressionFilter& lhs, const ExpressionFilter& rhs) {
-        return *(lhs.expression) == *(rhs.expression);
-    }
-};
-
-
-using FilterBase = variant<
-    class NullFilter,
-    class EqualsFilter,
-    class NotEqualsFilter,
-    class LessThanFilter,
-    class LessThanEqualsFilter,
-    class GreaterThanFilter,
-    class GreaterThanEqualsFilter,
-    class InFilter,
-    class NotInFilter,
-    class AnyFilter,
-    class AllFilter,
-    class NoneFilter,
-    class HasFilter,
-    class NotHasFilter,
-    class TypeEqualsFilter,
-    class TypeNotEqualsFilter,
-    class TypeInFilter,
-    class TypeNotInFilter,
-    class IdentifierEqualsFilter,
-    class IdentifierNotEqualsFilter,
-    class IdentifierInFilter,
-    class IdentifierNotInFilter,
-    class HasIdentifierFilter,
-    class NotHasIdentifierFilter,
-    class ExpressionFilter>;
-
-class Filter : public FilterBase {
-public:
-    using FilterBase::FilterBase;
+    Filter():expression({}) {}
```

Comment:
Is there an advantage to using `optional` instead of a `nullptr`? For better or for worse, `unique_ptr` has "optional" functionality built in and every callsite ought to handle it gracefully.